### PR TITLE
[PD-2541] implement nuo wizard history

### PR DESCRIPTION
**Pull request by koonsies**
## What's this PR do?

Adds history handling to the NUO Process Record step. This allows for use of back, forward, and reload when navigating the entire app.

## Where should the reviewer start?

If you have no NUO records, run this script **be sure to replace the email field**
```
%autoindent
from datetime import datetime
from factory import fuzzy
from myjobs.models import User
from mypartners.models import (OutreachRecord, OutreachEmailAddress,
                               OutreachWorkflowState)

user = User.objects.get(email="_____ENTER_YOUR_EMAIL_HERE_____")
company = user.roles.filter(name="Admin").first().company
try:
    outreach_email = OutreachEmailAddress.objects.get(company=company,
                                                      email="detestemail")
except:
    outreach_email = OutreachEmailAddress(company=company,
                                          email="detestemail")
    outreach_email.save()

try:
    workflow = OutreachWorkflowState.objects.get(state="New")
except:
    workflow = OutreachWorkflowState(state="New")
    workflow.save()

i = 0
while i < 100:
    i += 1
    orr = OutreachRecord(date_added=datetime.now(),
                   outreach_email=outreach_email,
                   from_email=str(fuzzy.FuzzyText(length=6).fuzz()) + "@example.com",
                   email_body=fuzzy.FuzzyText(length=200).fuzz(),
                   current_workflow_state=workflow
                   )
    orr.save()
    print orr.pk
```

And if you don't have -access- to NUO, run thiiiss script.
```
from myjobs.models import User
from seo.models import BusinessUnit
import datetime

aa = AppAccess.objects.get(name="nuo")
user = User.objects.get(email="ENTER EMAIL HERE")
company = user.roles.filter(name="Admin").first().company
company.app_access.add(aa)
```

## How should this be manually tested?
Head to NUO. (https://secure.my.jobs/prm/view/nonuseroutreach/)
Click 'Review' on one of the records.
Start doing the process steps. Navigate back and forward occasionally.
Make sure it works.
Let me know if it doesn't.

## Any background context you want to provide?

This PR also has a couple of refactors in it.. The most notable one was swapping state.process.state to state.process.step (state as a variable name is pretty bad). And that refactor changed code in a loott of places. Sorry.

### CODE REVIEW

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js [hunk 1]
```diff
@@ -234,6 +298,64 @@ function formsToApi(record, newTags) {
...
+}
+
+export function getErrorsForForms(forms, key) {
+  return map(get(forms, key), f => !isEmpty(get(f, 'errors')));
+}
+
+/**
+ *  Determine what the state of the process view should be
+ */
+export function doDetermineProcessState() {
```
Comment (koonsies) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: This isn't new code. I moved the logic from the reducer to an action. Cuz reducers ain't supposed to do dis much.

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js [hunk 2]
```diff
@@ -184,6 +214,40 @@ export function doLoadEmail(outreachId) {
...
+ */
+export function doEditContact(contactIndex) {
+  return (dispatch, getState) => {
+    const currentState = getState();
+    const pk = get(currentState, ['process', 'record', 'contacts', contactIndex, 'pk']);
+    const newStep = pk ? 'CONTACT_APPEND' : 'NEW_CONTACT';
+
+    // Ensure we do not repeat this call if not needed
+    if ((currentState.process.contactIndex !== contactIndex ||
+      currentState.process.step !== newStep) &
```
Comment (aljordan82) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Why bitwise operator is used?

Comment (koonsies) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Whoops

File: gulp/src/nonuseroutreach/actions/process-outreach-actions.js [hunk 3]
```diff
@@ -184,6 +214,40 @@ export function doLoadEmail(outreachId) {
...
+}
+
+/**
+ * determine if we need to change state to edit communication record form
+ */
+export function doEditCommunicationRecord() {
+  return (dispatch, getState) => {
+    const currentState = getState();
+    // Ensure we do not repeat this call if not needed
+    if (currentState.process.step !== 'NEW_COMMUNICATIONRECORD' &
```
Comment (aljordan82) [inline gulp/src/nonuseroutreach/actions/process-outreach-actions.js]: Another &

**Pull request closed without merging**